### ISSUE TITLE
Question: could PHPCS be extended to sniff Laravel's blade syntax?

### ISSUE THREAD

**Issue by u01jmg3** [squizlabs/PHP_CodeSniffer#437]
- Not an issue but a question regarding the appropriateness and extendibility of PHPCS and running its rules such as `ControlSignatureSniff` against Laravel's blade syntax?

**Comment by Konafets** [comment]
Does this answer your question? https://github.com/squizlabs/PHP_CodeSniffer/wiki/Coding-Standard-Tutorial

**Comment by u01jmg3** [comment]
- Not really - I already have a [custom ruleset](https://github.com/uofa/gulp-starter/blob/master/composer_plugins/phpcs-ruleset.xml) but wondered how I could utilise PHPCS to cover inconsistencies in blade syntax using rules like `ControlSignatureSniff`

**Comment by Konafets** [comment]
I dont know blade. Can you provide an code example of the inconsistencies?

**Comment by u01jmg3** [comment]
[Scroll down to `If Statements`](http://laravel.com/docs/4.2/templates#other-blade-control-structures) to see an example of blade syntax. More interested in spacing/layout issues which in PHP can be corrected with sniffs like `ControlSignatureSniff`.

**Comment by Konafets** [comment]
Do you know how PHPCS tokenizes the @ keywords? You can figure this out when you check the code with ```php -vv ...```. I assume its just string, than you need to implement everything by you own.

- Listen to every string
- Check if its one of the keywords (if, else, elseif, ...)
- Implememnt the violation check

**Comment by u01jmg3** [comment]
Thanks

Could you provide a snippet of a custom rule that would achieve this to get me started?

**Comment by Konafets** [comment]
Its tokenized to T_INLINE_HTML.

```
cat test.blade.php                                                                                                                               ⏎
<html>
    <body>
        @section('sidebar')
            This is the master sidebar.
        @show

        <div class="container">
            @yield('content')
        </div>
    </body>
</html>

@if (count($records) === 1)
    I have one record!
@elseif (count($records) > 1)
    I have multiple records!
@else
    I don't have any records!
@endif

@unless (Auth::check())
    You are not signed in.
@endunless
```

So you need to register that token:

```php
public function register()
{
    return array(
        T_INLINE_HTML,    
    );
}//end register()

...

public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr) {
    $tokens = $phpcsFile->getTokens();
    if ($tokens[$stackPtr]['content'] === '@if') {
        $endif = $phpcsFile->findNext(T_INLINE_HTML, $stackPtr + 1, null, false, '@endif');    
        // Now you have the scope opener and scope closer for this condition and are able to move between them.
   }
}

This approach only works for conditons which are not nested. Than you must implement detection which @ endif belongs to which @if and so on. PHPCS provides all this by scope opener/closer properties for native supported Tokenizers.

**Comment by u01jmg3** [comment]
Thanks for this - something I am going to have to invest some time in to cover what I would like to cover.

Do you think other Laravel users would find this useful?

**Comment by Konafets** [comment]
For me it seems your need to put too much effort into such checks. If it would help you or you need to implement such checks to help your collegues to meet the company CGL, it might worth it. Otherwise I don't know. I have no idea how big the community is and if the are interested into that. Such checks are worth nothing if there is no mindset of code quality behind. 

Only you knows if it makes sense, if ppls would use it and if you like to put effort into this.

**Comment by aik099** [comment]
Doesn't sound right. The Blade template engine, used in Laravel, uses it's own PHP-alike syntax. Because of this it's unlikely that any tools that are doing PHP tokenizing like PHP_CodeSniffer will be able to parse and validate it properly.

**Comment by gsherwood** [comment]
I think you'd need a custom tokenizer to really make the blade syntax easy to write sniffs for. You can write sniffs to check for inline HTML and parse it out that way, but you'll probably need to tokenize it in there anyway to make the code easy to check.

But that tokenizer is not going to produce something that the existing sniffs would be able to check and fix. So you'd also need to write custom sniffs to check every part of the blade syntax.

It is entirely possible to fork PHPCS and make it do this. But it's a heap of work and not something I could help with given my current workload.

**Comment by nmfzone** [comment]
In favor of Laravel Community, I think this feature is important. Hope it'll be implemented soon.

**Comment by MelchiorKokernoot** [comment]
> In favor of Laravel Community, I think this feature is important. Hope it'll be implemented soon.

Go ahead and propose a PR then!

**Comment by nmfzone** [comment]
> > In favor of Laravel Community, I think this feature is important. Hope it'll be implemented soon.
> 
> Go ahead and propose a PR then!

Aha, good idea. Maybe you can help me, Mr Achievement.

**Comment by MelchiorKokernoot** [comment]
Can you make your question a little more specific?
